ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,938 @@ PHP_FUNCTION(bcscale)
+) {
+	if (auto_scale) {
+		*scale = MAX(n1_full_scale, n2_full_scale);
+	}
+	if (!bc_modulo(n1, n2, ret, *scale)) {
+		zend_throw_exception_ex(zend_ce_division_by_zero_error, 0, "Modulo by zero");
+		return FAILURE;
+	}
+	bc_rm_trailing_zeros(*ret);
+	return SUCCESS;
+}
+
+static zend_always_inline zend_result bcmath_number_pow_internal(
+	bc_num n1, bc_num n2, bc_num *ret,
+	size_t n1_full_scale, size_t n2_full_scale, size_t *scale, bool auto_scale, bool is_op
+) {
+	/* Check the exponent for scale digits and convert to a long. */
+	if (UNEXPECTED(n2->n_scale != 0)) {
+		if (is_op) {
+			zend_value_error("exponent cannot have a fractional part");
+		} else{
+			zend_argument_value_error(1, "exponent cannot have a fractional part");
+		}
+		return FAILURE;
+	}
+	long exponent = bc_num2long(n2);
+
+	bool scale_expand = false;
+	if (auto_scale) {
+		if (exponent > 0) {
+			*scale = n1_full_scale * exponent;
+			if (UNEXPECTED(*scale > INT_MAX || *scale < n1_full_scale)) {
+				*scale = INT_MAX;
+			}
+		} else if (exponent < 0) {
+			*scale = MIN(n1_full_scale + BC_MATH_NUMBER_MAX_EX_SCALE, INT_MAX);
+			scale_expand = true;
+		} else {
+			*scale = 0;
+		}
+	}
+
+	if (UNEXPECTED(exponent == 0 && (n2->n_len > 1 || n2->n_value[0] != 0))) {
+		if (is_op) {
+			zend_value_error("exponent is too large");
+		} else {
+			zend_argument_value_error(1, "exponent is too large");
+		}
+		return FAILURE;
+	}
+	bc_raise(n1, exponent, ret, *scale);
+	bc_rm_trailing_zeros(*ret);
+	if (scale_expand) {
+		size_t diff = *scale - (*ret)->n_scale;
+		*scale -= diff > BC_MATH_NUMBER_MAX_EX_SCALE ? BC_MATH_NUMBER_MAX_EX_SCALE : diff;
+	}
+	return SUCCESS;
+}
+
+static zend_always_inline bcmath_number_obj_t *bcmath_number_new_obj(bc_num ret, size_t scale)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(bcmath_number_create(bcmath_number_ce));
+	intern->num = ret;
+	intern->scale = scale;
+	return intern;
+}
+
+static zend_always_inline zend_result bcmath_number_parse_num(zval *zv, zend_object **obj, zend_string **str, zend_long *lval)
+{
+	if (Z_TYPE_P(zv) == IS_OBJECT && instanceof_function(Z_OBJCE_P(zv), bcmath_number_ce)) {
+		*obj = Z_OBJ_P(zv);
+		return SUCCESS;
+	} else {
+		switch (Z_TYPE_P(zv)) {
+			case IS_LONG:
+				*lval = Z_LVAL_P(zv);
+				return SUCCESS;
+
+			case IS_STRING:
+				*str = Z_STR_P(zv);
+				return SUCCESS;
+
+			default:
+				return zend_parse_arg_str_or_long_slow(zv, str, lval, 1 /* dummy */) ? SUCCESS : FAILURE;
+		}
+	}
+}
+
+static zend_always_inline void bc_num_from_zend_object(const zend_object *obj, bc_num *num, size_t *full_scale)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	*num = intern->num;
+	if (full_scale) {
+		*full_scale = intern->scale;
+	}
+}
+
+static zend_always_inline zend_result bc_num_from_zend_string(const zend_string *str, bc_num *num, size_t *full_scale)
+{
+	if (UNEXPECTED(php_str2num_ex(num, str, full_scale) == FAILURE)) {
+		bc_free_num(num);
```

Comment:
I'm not sure why you free on failure here, if the function fails then num isn't initialized with a number right?

Reply:
Oh, right

Reply:
Fixed in https://github.com/php/php-src/pull/13741/commits/c30216787240bb574646fd6d43e4cb2eb00e3d91

Reply:
At the time of failure, the existing code was copying `BCG(_ZERO_)`.
If don't free this up, it will leak. Therefore, I reverted the changes.

https://github.com/php/php-src/pull/13741/commits/22b08ecc5a2f753e6949c2c84c5cda082f69ad00